### PR TITLE
INTMDB-222: Added Synonyms to Search Index RS and DS

### PR DESCRIPTION
**Pull request by abner-dou**
## Description

Added synonyms support to search index resource and datasource

Link to any related issue(s):[INTMDB-222](https://jira.mongodb.org/browse/INTMDB-222)

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [x] Documentation fix/enhancement

## Required Checklist:

- [ ] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [ ] I have read the Terraform contribution guidelines
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code

### CODE REVIEW

File: mongodbatlas/resource_mongodbatlas_search_index.go [hunk 1]
```diff
@@ -322,6 +367,28 @@ func resourceMongoDBAtlasSearchIndexCreate(ctx context.Context, d *schema.Resour
...
+			synonymsDoc["name"] = synonym["name"]
+			synonymsDoc["analyzer"] = synonym["analyzer"]
+			synonymsDoc["source"] = map[string]interface{}{
+				"collection": synonym["source_collection"],
+			}
+			synonymsList = append(synonymsList, synonymsDoc)
+		}
+	}
+	return synonymsList
+}
```
Comment (andreaangiolillo) [inline mongodbatlas/resource_mongodbatlas_search_index.go]:
```suggestion
func expandSearchIndexSynonyms(d *schema.ResourceData) []map[string]interface{} {
	var synonymsList []map[string]interface{}

	synonymsDoc := map[string]interface{}{}

	if vSynonyms, vSynonymsOK := d.GetOk("synonyms"); vSynonymsOK{
		for _, s := range vSynonyms.(*schema.Set).List() {
			synonym := s.(map[string]interface{})

			synonymsDoc["name"] = synonym["name"]
			synonymsDoc["analyzer"] = synonym["analyzer"]
			synonymsDoc["source"] = map[string]interface{}{
				"collection": synonym["source_collection"],
			}
			synonymsList = append(synonymsList, synonymsDoc)
		}
	}
	
	return synonymsList
}
```

**Review by andreaangiolillo — APPROVED** [review]
LGTM

**Review by thetonymaster — APPROVED** [review]
LGTM just check @andreaangiolillo's comment

File: mongodbatlas/resource_mongodbatlas_search_index.go (hunk 1 above)
Comment (abner-dou) [reply mongodbatlas/resource_mongodbatlas_search_index.go]: already changed :D

**Review by themantissa — APPROVED** [review]
LGTM but just pinged the search team to make sure okay with them.  Standby for their input.

**Review by navenoxin — CHANGES_REQUESTED** [review]
Am I reading this correctly in that we are changing the syntax slightly here? Can we use the original syntax where a synonym mapping looks like

```json5
{
  name: "synonym mapping name",
  source: {
    collection: "source collection"
  },
  analyzer: "some analyzer"
}
```

instead of

```json5
{
  name: "synonym mapping name",
  source_collection: "source collection",
  analyzer: "some analyzer"
}
```

We do not want to change the index definition syntax from what is presented elsewhere and [explained in the docs](https://docs.atlas.mongodb.com/reference/atlas-search/synonyms/#syntax).

**Comment by abner-dou** [comment]
@navenoxin In response for your latest comment:

Are you planning to add more properties to the `source` document or just the `collection` properties, because adding sub-levels to the properties adds more complexity to the resource in terraform provider.

**Comment by navenoxin** [comment]
@abner-dou We do not plan on adding more properties to the source document in the immediate future. Understood re: this adding a little complexity to the terraform provider, but I'd argue it is more important to present a uniform syntax everywhere this is exposed.

CC @themantissa @doug-ol

File: website/docs/d/search_index.html.markdown [hunk 2]
```diff
@@ -35,11 +35,15 @@ data "mongodbatlas_search_index" "test" {
 * `name` - Name of the index.
 * `analyzer` - [Analyzer](https://docs.atlas.mongodb.com/reference/atlas-search/analyzers/#std-label-analyzers-ref) to use when creating the index.
 * `analyzers` - [Custom analyzers](https://docs.atlas.mongodb.com/reference/atlas-search/analyzers/custom/#std-label-custom-analyzers) to use in this index (this is an array of objects).
-* `collection_name` - (Required) Name of the collection the index is on.
-* `database` - (Required) Name of the database the collection is in.
+* `collection_name` - Name of the collection the index is on.
+* `database` - Name of the database the collection is in.
```
Comment (navenoxin) [inline website/docs/d/search_index.html.markdown:39]: why remove `(Required)` here?

File: website/docs/d/search_index.html.markdown [hunk 3]
```diff
@@ -35,11 +35,15 @@ data "mongodbatlas_search_index" "test" {
...
 * `mappings_dynamic` - Flag indicating whether the index uses dynamic or static mappings.
 * `mappings_fields` - Object containing one or more field specifications.
 * `search_analyzer` - [Analyzer](https://docs.atlas.mongodb.com/reference/atlas-search/analyzers/#std-label-analyzers-ref) to use when searching the index.
+* `synonyms` - 	Synonyms mapping definition to use in this index.
+* `synonyms.#.name` - Name of the [synonym mapping definition](https://docs.atlas.mongodb.com/reference/atlas-search/synonyms/#std-label-synonyms-ref).
+* `synonyms.#.source_collection` - Name of the source MongoDB collection for the synonyms.
+* `synonyms.#.analyzer` - Name of the [analyzer](https://docs.atlas.mongodb.com/reference/atlas-search/analyzers/#std-label-analyzers-ref) to use with this synonym mapping. If `mappings.dynamic` is `true`, Atlas uses the default analyzer, [lucene.standard](https://docs.atlas.mongodb.com/reference/atlas-search/analyzers/standard/#std-label-ref-standard-analyzer), for synonym mapping.
```
Comment (navenoxin) [inline website/docs/d/search_index.html.markdown]: please remove this, it is not true.

```suggestion
* `synonyms.#.analyzer` - Name of the [analyzer](https://docs.atlas.mongodb.com/reference/atlas-search/analyzers/#std-label-analyzers-ref) to use with this synonym mapping.
```

File: website/docs/d/search_indexes.html.markdown [hunk 4]
```diff
@@ -51,6 +51,11 @@ data "mongodbatlas_search_index" "test" {
 * `mappings_dynamic` - Flag indicating whether the index uses dynamic or static mappings.
 * `mappings_fields` - Object containing one or more field specifications.
 * `search_analyzer` - [Analyzer](https://docs.atlas.mongodb.com/reference/atlas-search/analyzers/#std-label-analyzers-ref) to use when searching the index.
+* `synonyms` - 	Synonyms mapping definition to use in this index.
+* `synonyms.#.name` - Name of the [synonym mapping definition](https://docs.atlas.mongodb.com/reference/atlas-search/synonyms/#std-label-synonyms-ref).
+* `synonyms.#.source_collection` - Name of the source MongoDB collection for the synonyms.
+* `synonyms.#.analyzer` - Name of the [analyzer](https://docs.atlas.mongodb.com/reference/atlas-search/analyzers/#std-label-analyzers-ref) to use with this synonym mapping. If `mappings.dynamic` is `true`, Atlas uses the default analyzer, [lucene.standard](https://docs.atlas.mongodb.com/reference/atlas-search/analyzers/standard/#std-label-ref-standard-analyzer), for synonym mapping.
```
Comment (navenoxin) [inline website/docs/d/search_indexes.html.markdown]: please remove this, it is not true.

```suggestion
* `synonyms.#.analyzer` - Name of the [analyzer](https://docs.atlas.mongodb.com/reference/atlas-search/analyzers/#std-label-analyzers-ref) to use with this synonym mapping.
```

File: website/docs/r/search_index.html.markdown [hunk 5]
```diff
@@ -345,13 +351,33 @@ An [Atlas Search analyzer](https://docs.atlas.mongodb.com/reference/atlas-search
...
+* `source_collection` - (Required) Name of the source MongoDB collection for the synonyms. Documents in this collection must be in the format described in the [Synonyms Source Collection Documents](https://docs.atlas.mongodb.com/reference/atlas-search/synonyms/#std-label-synonyms-coll-spec).
+* `analyzer` - (Required) Name of the [analyzer](https://docs.atlas.mongodb.com/reference/atlas-search/analyzers/#std-label-analyzers-ref) to use with this synonym mapping. If you set `mappings.dynamic` to `true`, you must use the default analyzer, [lucene.standard](https://docs.atlas.mongodb.com/reference/atlas-search/analyzers/standard/#std-label-ref-standard-analyzer), here also. Atlas Search doesn't support these [custom analyzer](https://docs.atlas.mongodb.com/reference/atlas-search/analyzers/custom/#std-label-custom-analyzers) tokenizers and token filters in the index definition for [synonyms](https://docs.atlas.mongodb.com/reference/atlas-search/synonyms/#std-label-synonyms-ref):
```
Comment (navenoxin) [inline website/docs/r/search_index.html.markdown]: changed note re: dynamic indexes, clarified scope of where custom analyzer components are invalid

```suggestion
* `analyzer` - (Required) Name of the [analyzer](https://docs.atlas.mongodb.com/reference/atlas-search/analyzers/#std-label-analyzers-ref) to use with this synonym mapping. Atlas Search doesn't support these [custom analyzer](https://docs.atlas.mongodb.com/reference/atlas-search/analyzers/custom/#std-label-custom-analyzers) tokenizers and token filters in [analyzers used in synonym mappings](https://docs.atlas.mongodb.com/reference/atlas-search/synonyms/#options):
```

**Review by navenoxin — CHANGES_REQUESTED** [review]
A few corrections, otherwise no blocking objections

**Review by navenoxin — CHANGES_REQUESTED** [review]
@abner-dou please address suggestions from my last review cc @themantissa

**Review by navenoxin — APPROVED** [review]

**Pull request merged**